File: python/avoid_negative_conditionals.py
# Avoid Negative Conditionals
# Positive conditionals reduce mental strain and make it easier to reason about the code.

# Before
class User:
    def __init__(self):
        self.is_admin = False
        self.is_premium = False

    def admin(self):
        return self.is_admin

    def premium_member(self):
        return self.is_premium

def check_access(user):
    if not user.admin():
        if not user.premium_member():
            print("Access denied.")
        else:
            print("Access granted.")
    else:
        print("Access granted.")

user = User()
check_access(user)


# After
class User:
    def __init__(self):
        self.is_admin = False
        self.is_premium = False

    def admin(self):
        return self.is_admin

    def premium_member(self):
        return self.is_premium

def check_access(user):
    if user.admin() or user.premium_member():
        print("Access granted.")
    else:
        print("Access denied.")

user = User()
check_access(user)

File: javascript/avoid_negative_conditionals.js
// Avoid Negative Conditionals
// Positive conditionals reduce mental strain and make it easier to reason about the code.

// Before
class User {
  constructor() {
    this.isAdmin = false;
    this.isPremium = false;
  }

  admin() {
    return this.isAdmin;
  }

  premiumMember() {
    return this.isPremium;
  }
}

function checkAccess(user) {
  if (!user.admin()) {
    if (!user.premiumMember()) {
      console.log("Access denied.");
    } else {
      console.log("Access granted.");
    }
  } else {
    console.log("Access granted.");
  }
}

const user = new User();
checkAccess(user);


// After
class User {
  constructor() {
    this.isAdmin = false;
    this.isPremium = false;
  }

  admin() {
    return this.isAdmin;
  }

  premiumMember() {
    return this.isPremium;
  }
}

function checkAccess(user) {
  if (user.admin() || user.premiumMember()) {
    console.log("Access granted.");
  } else {
    console.log("Access denied.");
  }
}

const user = new User();
checkAccess(user);
